### ISSUE TITLE
Using AzureServiceBus with azure functions

### ISSUE THREAD

**Issue by Tsjunne** [rebus-org/Rebus.AzureServiceBus#68]
I was browsing the web for some clues on how to effectively use Rebus to host message handlers in an Azure Function.
There are many opportunities for cost savings on Azure when moving to serverless functions, but I hate to loose the added value of the Rebus pipeline.
I fell upon an implementation for NServiceBus:
https://docs.particular.net/samples/azure-functions/
So this is clearly possible from Azure Extensibility view.

Has this been done or tried before using Rebus? Is this even remotely possible in Rebus' current architecture?

**Comment by mookid8000** [comment]
Yes I think it would be possible, but there's no API that helps with configuring it at the moment.

Rebus receives messages using "workers", and then incoming messages are dispatched by calling a "pipeline invoker" – you can see the lines responsible [here](https://github.com/rebus-org/Rebus/blob/master/Rebus/Workers/ThreadPoolBased/ThreadPoolWorker.cs#L165-L196). 

As you can probably see, the design fully supports the "here, Rebus, take this message and handle it" scenario 🙂

**Comment by georgechond94** [comment]
Hi @mookid8000, any guidelines on how we can achieve this?
I tried to still use the `ServiceBusTrigger` in my Function and then dispatch the message to the Rebus pipeline, but I had no luck with it yet.

Thanks!

**Comment by mookid8000** [comment]
Can you tell me how far you got?

I know that e.g. [`DefaultPipelineInvoker`](https://github.com/rebus-org/Rebus/blob/master/Rebus/Pipeline/Invokers/DefaultPipelineInvoker.cs) is internal (and possibly other related classes), so unfortunately it's not just
```csharp
using var scope = new RebusTransactionScope();
var transportMessage = (...);
var invoker = new DefaultPipelineInvoker(...);
var context = new IncomingStepContext(transportMessage, scope.TransactionContext);
await invoker.Invoke(context);
await scope.CompleteAsync();
```
buuut I don't think you would need to copy that many lines of code to emulate that in a function....

**Comment by michielvoo** [comment]
I have not found a way to create a Rebus pipeline and send the message through it. Many classes required to build a pipeline are internal (for example the JSON serializers), and I believe this is because the Rebus' fluent configuration API hides the actual classes and only exposes extensions methods (for example `UseNewtonsoftJson()`). 

So it seems that to support this use case would require opening up some parts of the Rebus API to allow one to easily create a fully functional pipeline without an actual bus.

**Comment by C0DK** [comment]
So is this still an open issue? :) It doesn't seem to be possible. Or am I skipping relevant parts of the discussion?
